### ISSUE TITLE
[grpc-c++] grpc::GoogleDefaultCredentials() segfaults on macOS and FreeBSD

### ISSUE THREAD

**Issue by bdhess** [grpc/grpc#28764]
### What version of gRPC and what language are you using?
gRPC at HEAD
From running `git bisect` it looks like the problem was introduced at commit d1297bcc173da4d83926bb72bfd98d1984bb4582

### What operating system (Linux, Windows,...) and version?
macOS 10.14
FreeBSD 11
(Nb the test below passes on Linux and Windows)

### What runtime / compiler are you using (e.g. python version or version of gcc)
Clang - the versions that ship with macOS 10.14 and FreeBSD 11

### What did you do?
```
#include <memory>

#include "grpcpp/security/credentials.h"
#include "gmock/gmock.h"
#include "gtest/gtest.h"

namespace {

using ::testing::IsNull;
using ::testing::Not;

TEST(GrpcSegvTest, DemonstrateSegv) {
    std::shared_ptr<grpc::ChannelCredentials> creds = grpc::GoogleDefaultCredentials();
    ASSERT_THAT(creds, Not(IsNull()));
}

}
```

### What did you expect to see?
Test passes

### What did you see instead?
Test segfaults

### Anything else we should know about your project / environment?

On macOS, Google credentials are provided with the GOOGLE_APPLICATION_CREDENTIALS environment variable.
On FreeBSD, system default credentials from GCE are used.

**Comment by bdhess** [comment]
@veblush fyi

**Comment by bdhess** [comment]
Sorry now that I'm looking at this more closely it looks like this is actually fixed at HEAD but not in 1.43.2

**Comment by bdhess** [comment]
Ok confirmed this is fixed at 9f1663fb74 -- which I had to find the hard way because bisect didn't work because merges, grumble.

It seems like the fix was sort of inadvertent, so leaving this issue open to track fixing the testing gap.

**Comment by bdhess** [comment]
This is broken again in v1.44.0 and at head.

Tracking these down with git bisect is hard because of the way these get merged, but here's where we are:

```
[bdhess@buildchain-freebsd /opt/grpc]$ git bisect bad
The merge base 6b95e9769e241afc43e8ae70b24bdac49570ee59 is bad.
This means the bug has been fixed between 6b95e9769e241afc43e8ae70b24bdac49570ee59 and [6bf8e2248444ea331299e80a302f67cbf1de41b5].
```

**Comment by veblush** [comment]
@bdhess gRPC has a similar test [here](https://github.com/grpc/grpc/blob/2223877d417ca1a6fd06db7dc585bd1d35e540a0/test/cpp/client/credentials_test.cc#L66) but it doesn't check whether it's nullptr or not. The segfault you saw came from `ASSERT_THAT(creds, Not(IsNull()));` or somewhere gRPC internal? If it's coming from the internal, stacktrace would be helpful. Also I'd note that FreeBSD isn't a supported platform so it might have some issues.

**Comment by bdhess** [comment]
Ah sorry, yes, to be clear, the segfault is somewhere within the call to `grpc::GoogleDefaultCredentials()`.

It looks to be fixed again at head but is repro-able at v1.44.0.

Wdym by stack trace -- setting the value of `GRPC_TRACE` to something and capturing stderr?  Or is there some other way of building/instrumenting.  Ftr, we are building with Bazel.

**Comment by bdhess** [comment]
Ok, I think the issue here is the default SSL certs not being loadable on those platforms (and the stderr output from grpc not being super visible in our tests).

tdbhacks and I are going to follow up separately.
